lib/index.ts

```diff
@@ -21,116 +21,105 @@ limitations under the License.
-		 * @method
-		 * @private
-		 *
-		 * @param {Object} options - request options
-		 * @returns {Promise<*>} response body
-		 *
-		 * @todo Implement caching support.
-		 */
-		public _request(
-			options: {
-				method: string;
-				url: string;
-				body?: PinejsClientCoreFactory.AnyObject;
-			} & PinejsClientCoreFactory.AnyObject,
-		) {
-			return auth.hasKey().then(function(hasKey) {
-				const authenticated = hasKey || (apiKey != null && apiKey.length > 0);
-
-				options = Object.assign(
-					{
-						apiKey,
-						baseUrl: apiUrl,
-						sendToken: authenticated && !options.anonymous,
-					},
-					options,
-				);
-
-				return request
-					.send(options)
-					.get('body')
-					.catch(function(err) {
-						if (err.statusCode !== 401) {
-							throw err;
-						}
-
-						// Always return the API error when the anonymous flag is used.
-						if (options.anonymous) {
-							throw err;
-						}
-
-						// We want to allow unauthenticated users to make requests
-						// to public resources, but still reject with a NotLoggedIn
-						// error if the response ends up being a 401.
-						if (!authenticated) {
-							throw new errors.BalenaNotLoggedIn();
-						}
-
-						throw err;
-					});
-			});
+	public async _request(
+		options: {
+			method: string;
+			url: string;
+			body?: AnyObject;
+		} & AnyObject,
+	) {
+		const { apiKey, apiUrl, auth, request } = this.backendParams;
+
+		const hasKey = await auth.hasKey();
+		const authenticated = hasKey || (apiKey != null && apiKey.length > 0);
+
+		options = {
+			apiKey,
+			baseUrl: apiUrl,
+			sendToken: authenticated && !options.anonymous,
+			...options,
+		};
+
+		try {
+			const { body } = await request.send(options);
+			return body;
+		} catch (err) {
+			if (err.statusCode !== 401) {
+				throw err;
+			}
+
+			// Always return the API error when the anonymous flag is used.
+			if (options.anonymous) {
+				throw err;
+			}
+
+			// We want to allow unauthenticated users to make requests
+			// to public resources, but still reject with a NotLoggedIn
+			// error if the response ends up being a 401.
+			if (!authenticated) {
+				throw new errors.BalenaNotLoggedIn();
+			}
+
+			throw err;
 		}
 	}
-
-	const pineInstance = new BalenaPine({
-		apiPrefix,
-	});
-
-	return Object.assign(pineInstance, {
-		API_URL: apiUrl,
-		API_VERSION: apiVersion,
-		API_PREFIX: apiPrefix,
-	});
 }
 
-// tslint:disable-next-line:no-namespace
-declare namespace getPine {
-	// We have to declare this in a namespace to avoid an error around using private types
-	// in declaration files
-	export class BalenaPine extends PinejsClientCoreFactory.PinejsClientCore<
-		BalenaPine
-	> {
-		public _request(
-			options: {
-				method: string;
-				url: string;
-				body?: PinejsClientCoreFactory.AnyObject;
-			} & PinejsClientCoreFactory.AnyObject,
-		): Promise<any>;
-	}
-}
+export function getPine(params: BackendParams) {
+	const pineInstance = new BalenaPine({}, params);
```

Comment:
You could replace this by doing `export class BalenaPine ...` and letting the user call `new` - it'd also allow extending the class if there was reason to do so

Reply:
Can do, but having the first argument empty made me think to keep the easier to digest factory method.
How about keeping both?

Reply:
Well, we have README & typings to explain this... Yea, will do.